[gatsby-plugin-offline] [gatsby/cache-dir] Fix various offline and caching issues

davidbailey00 opened the pull request:
Changes made:

- `<div>` is replaced with `<span>` in gatsby-plugin-offline's `app-shell.js` to prevent an edge case in Chrome, where the the outer element's class attribute is unset despite its className being set in the VDOM, if the page was loaded from the offline shell.
- A call to `getResourcesForPathname` was fixed - previously this prevented later code from calling since `Promise.all(pathnameResources)` would never resolve.
- Styles are now searched as well as scripts when caching resources, and these are all cached regardless of whether they are included in `pageResources`.
- Styles are also now searched in `get-resources-from-html.js`.
- The offline shell and custom 404 page are now also scanned for resources when considering critical file paths.
  - `get-resources-from-html.js` is modified so that if a custom 404 page doesn't exist, an empty array will be returned.
- A glob for Webpack chunks is added to the precache options, as these are sometimes loaded from other chunks and therefore not detected by `getResourcesFromHTML`, causing errors to occur when trying to navigate to another page when offline (if they haven't all been cached).
- The `navigateFallbackWhitelist` regex is modified to prevent loading the offline shell if the URL ends with `?no-cache=1` or `&no-cache=1`.
- `sw.js` is patched to include search queries when matching URLs against `navigateFallbackWhitelist`. Unfortunately I couldn't find any options to enable this, so patching appears to be the only way until a change is made upstream.
  - An error is thrown asking the user to report an issue if the patch fails - this will cause the build to fail, but will avoid confusion in the long run if the patch stops working.
  - I checked the `ignoreUrlParametersMatching` options, but it's for ignoring differences, e.g. between `/page/` and `/page/?utm_source=github`, not for disabling caching or fallback.
- JSON and CSS resources are cached at runtime.
- `page-renderer.js` now checks for `this.state.pageResources.json`, since errors were previously occurring even when `this.state.pageResources` existed due to the missing JSON data.
- When a page isn't found, an attempt is made to load it directly, in order to display a proper error message, i.e. 404 or offline error, or simply load it directly if it's available online (e.g. `/admin/` for CMS, which is not directly part of Gatsby).
- A clause has been added to an if statement in `production-app.js` to prevent the 404 showing until an attempt has been made to load the page directly
  - #7268 had actually broken support for static pages (e.g. Netlify CMS at /admin/) since the service worker would be unable to find the page from Gatsby and show a 404
- `onServiceWorkerInstalled` is now delayed by 100ms, since it was being called before the `fetch` event listeners were added, thus preventing resources from being cached. Unfortunately I couldn't find a programmatic way to do this (other than more patching of `sw.js`, which is less than ideal).
  - **It might be worth increasing this to 500ms, since at 100ms it could still fail for more complex sites (despite working at as low as 20ms in local testing). Thoughts?**
- `checkIfDoingInitialRenderForSW` is removed from `loader.js` as it was causing the SW to be removed when navigating to a nonexistent page, even if the website had not been updated, causing it to stop working offline *entirely* - regardless, it is no longer needed with the other changes made.
- README for gatsby-plugin-offline is updated to match the latest default options.

Closes #6212, Closes #7053, Closes #6688, possibly others

KyleAMathews commented:
Deploy preview for *using-postcss-sass* failed.

Built with commit 15ee19294647dbcbb7861507d5d7ccb80211645f

https://app.netlify.com/sites/using-postcss-sass/deploys/5b745f3b4ed62f238b914390

KyleAMathews commented on packages/gatsby-plugin-offline/package.json:
```diff
@@ -1,7 +1,7 @@
   "name": "gatsby-plugin-offline",
   "description": "Gatsby plugin which sets up a site to be able to run offline",
-  "version": "2.0.0-beta.9",
+  "version": "2.0.0-beta.10",
```
the version gets bumped automatically as part of deploys so please remove this

KyleAMathews commented:
Deploy preview for *image-processing* failed.

Built with commit 51c2e101493291b0e08789b16563013352272e67

https://app.netlify.com/sites/image-processing/deploys/5b74612182d3f155ed309d2c

davidbailey00 commented:
@KyleAMathews thanks! I'd normally agree with small PRs, but the bugs were all so interdependent that I felt they would become fragmented with separate PRs, which might result in some of them reappearing after others having been fixed or some being forgotten. Looking forward to your full review though!

KyleAMathews commented:
Deploy preview for *gatsbyjs* failed.

Built with commit 15ee19294647dbcbb7861507d5d7ccb80211645f

https://app.netlify.com/sites/gatsbyjs/deploys/5b745f394ed62f238b914384

KyleAMathews commented:
Deploy preview for *gatsbyjs* failed.

Built with commit 51c2e101493291b0e08789b16563013352272e67

https://app.netlify.com/sites/gatsbyjs/deploys/5b74611f82d3f155ed309d1a

davidbailey00 commented:
By the way, I just checked #7268 after rereading your comments on #6212, and it's a different issue to the one I was fixing -- #7268 serves the 404 page when running `gatsby serve`, whereas this PR is about handling what happens when the offline shell can't load the specified page, since the shell would previously try to handle all pages matching the regex `/^.*([^.]{5}|.html)$/` regardless of what's happening on the server.

wKovacs64 commented:
@davidbailey00: Nice! ❤️ 

@jlengstorf: A lot going on here, but some of it may address some or even all of the things pertaining to `gatsby-plugin-offline` that we went over in our pairing session last week. (Did you end up creating that issue somewhere? We may want to link here if so.)

jlengstorf commented:
@wKovacs64 I... definitely did not create that issue. I still have it written in my notebook to capture. 😬 

Thanks for the heads up!

wKovacs64 commented:
@jlengstorf Oh no sweat, no rush. Thanks!

davidbailey00 commented:
I've fixed the remaining issues with regards to 404 pages, including a bug which #7268 introduced, and I've updated the changelog to reflect these fixes. Code should now be ready for a review! :smile:

KyleAMathews commented:
Not necessarily for this PR but it'd be great to add this in the future https://github.com/gatsbyjs/gatsby/issues/7382

KyleAMathews commented:
Nice! Appreciate your thoroughness on this PR! It's definitely a tricky thing to get right.

This is our existing cypress test for `gatsby develop` on gatsbygram https://github.com/gatsbyjs/gatsby/tree/master/examples/gatsbygram/cypress/integration

davidbailey00 commented:
After some testing I found that when *not* using the offline plugin, 404 pages would unnecessarily redirect to `?no-cache=1` (this is only useful when using a service worker). After rewriting some code to make it more efficient, #7371 occurs again - it looks like I'd only covered up this problem and not actually fixed it.

However, this shouldn't be too difficult to fix properly - I'm just working on it now and should be done later this evening.

davidbailey00 commented:
Might have to work on this some more tomorrow - there are still some issues, and various bugs (re)appearing after I fix others. Having said that, I've made a lot of progress overall, and I think these are some of the last few bugs to squash!

KyleAMathews commented:
Great!

pieh commented on packages/gatsby/cache-dir/loader.js:
```diff
@@ -352,6 +319,22 @@ const queue = {
+          if (window.location.search) {
+            window.location.search += `&no-cache=1`
+          } else {
+            window.location.search = `?no-cache=1`
```
We use `getResourcesForPathname` not only for current path (for example it's used in `PageRenderer` that is used to render page in background in some views (gatsbygram, showcase on gatsbyjs.org) and thisi might cause weird issues potentially - should this be moved to navigation code if we can't get resources?

pieh replied:
also used in `hovering` action - https://github.com/gatsbyjs/gatsby/blob/3fd71b036343c451a3cf4901b4e5c991fd154451/packages/gatsby/cache-dir/loader.js#L196 - it would be not great UX if we hover link for path we can't get resources and browser refreshes automatically?

davidbailey00 replied:
That's true - I actually noticed this slightly earlier and fixed it in my local code, but I haven't pushed it yet since I'm still trying to get everything to work together properly and fix the last few bugs.

pieh commented on packages/gatsby/cache-dir/page-renderer.js:
```diff
@@ -119,7 +119,8 @@ class PageRenderer extends React.Component {
 
   render() {
-    if (!this.state.pageResources) return null
+    if (!(this.state.pageResources && this.state.pageResources.json))
```
was this tested in `develop`? I think in develop data comes from props passed by `json-store.js` and not from state

davidbailey00 replied:
I haven't explicitly tested this in develop yet - however, I think it should be fine since it was working previously with just the `!this.state.pageResources` check. I'll test properly once I've fixed the problems I'm currently working on.

davidbailey00 commented:
I think everything's almost ready now - just checking everything before I push the latest commit

![screenshot from 2018-08-16 23-28-04](https://user-images.githubusercontent.com/4248177/44238323-1eeb8b00-a1ac-11e8-868a-c13ab4711ee0.png)

azdanov commented:
@davidbailey00 By any chance is #7254 fixed by this PR?

davidbailey00 commented:
Happy to help! Having said that, some of the changes I've made in this PR might have made the transition smoother (even with the page refresh) since the offline shell and various other components are now cached properly.

davidbailey00 commented:
I've completed all 32 manual tests, all of which are now passing - looks like everything is now working properly! Ready for another review :smile: 

![screenshot from 2018-08-17 09-15-00](https://user-images.githubusercontent.com/4248177/44255464-1cb21c80-a1fe-11e8-97b6-6287e7477e09.png)

davidbailey00 commented:
Ah nope, another thing I forgot to do... `gatsby develop` wasn't tested separately with/without custom 404 pages, and I think they stopped working with some changes I made. Another commit incoming - I'm SURE this will be the last fix!!

davidbailey00 commented:
Might have to work some more on this - it looks like the changes to `serve.js` in #7268 only worked because of lack of proper caching, which of course is now fixed. I can confirm this is one of the last few problems - all of the problems with offline support and caching which I encountered are now fixed, and my test sites are working a lot more smoothly with a variety of different configurations.

davidbailey00 commented:
I think I've finally fixed the remaining 404 issues with gatsby dev - just going to run through all the tests again to check I haven't broken anything else!

KyleAMathews commented:
Sweet!!

KyleAMathews commented:
Random thought not associated with this PR. I was thinking about building into plugin-offline an auto-expiration of the SW cache (for index.html and the app shell) e.g. two weeks so that it someone hasn't visited a site for a while and is online, they see a new version instead of the old version. I think for content sites that's generally what you want. A news site could even set this to hours.

davidbailey00 commented:
There are a few issues due to some additional optimisations I made earlier (which I haven't pushed yet) - just going to fix those, but it should be fairly easy compared to everything else so far since I know what I changed. Once all the tests have passed I'll try to restrain myself from committing anything else in this PR!!

davidbailey00 commented:
Ah only just saw the second message @KyleAMathews (hadn't refreshed GitHub) - IMO it would be better to check if a new version has been deployed on page load and reload if that's the case. (In fact, I think this might already be the current behaviour?) This has all the benefits of your proposal (i.e. users always see the latest version) but without causing it to stop working offline after an arbitrary amount of time.

KyleAMathews commented:
It'd always work offline if the user is offline. But yeah, your proposal is interesting. Perhaps the offline plugin could write out a version to a file and then the SW could check for that file. If there's a new version it does a forced reload, if not, it serves from the cache.

KyleAMathews commented:
There's difficult tradeoffs though as what if the user is on a poor network. It might be better to serve an old version quickly instead of waiting. Perhaps there'd be a timeout though before serving from the cache.

KyleAMathews commented:
Or perhaps the head of index.html could have a version so we always start fetching that immediately and then the SW could read the first bytes and decide to keep going with it or serve from the cache.

davidbailey00 commented:
Just to clarify, I pushed the new file so everything is now ready, and I'm going to hold off making any further commits to this PR (unless I find one of the above tests sometimes fails).

KyleAMathews commented on packages/gatsby-plugin-offline/README.md:
```diff
@@ -29,34 +29,42 @@ and AppCache setup by changing these options so tread carefully.
-  ],
+    // Webpack chunks aren't always detected by `getResourcesFromHTML`
+    // since they're sometimes loaded from other chunks.
+    `${rootDir}/+([1234567890])-*.js`,
```
We don't want to precache every single webpack chunk — this could lead to precaching megabytes of JS on larger sites.

davidbailey00 replied:
Ah right - I had some offline issues (blank pages) when these files weren't being cached, but if there are any problems after I've removed this line I'll try and find another way around them.

davidbailey00 replied:
Good news, I'm no longer having problems after removing this line! No further changes needed

KyleAMathews commented on packages/gatsby-plugin-offline/README.md:
```diff
@@ -29,34 +29,42 @@ and AppCache setup by changing these options so tread carefully.
-      // Add runtime caching of images.
-      urlPattern: /\.(?:png|jpg|jpeg|webp|svg|gif|tiff)$/,
+      // Add runtime caching of various page resources.
+      urlPattern: /\.(?:png|jpg|jpeg|webp|svg|gif|tiff|js|woff|woff2|json|css)$/,
```
ugh, we weren't cache json files (page data) 🤦‍♂️

KyleAMathews commented on packages/gatsby-plugin-offline/README.md:
```diff
@@ -29,34 +29,42 @@ and AppCache setup by changing these options so tread carefully.
+  replacePrefix: args.pathPrefix || ``,
   navigateFallback: `/offline-plugin-app-shell-fallback/index.html`,
-  // Only match URLs without extensions.
+  // Only match URLs without extensions or the query `no-cache=1`.
```
where's this "no-cache=1" coming from?

KyleAMathews replied:
oh, I see I think. You're using it as a way to to ensure the sw does catch the reload?

davidbailey00 replied:
It's for when Gatsby detects a page not found after loading from the offline shell - usually it's correct, but in some cases the website could've been updated, or it's a static file which isn't handled by Gatsby (e.g. Netlify CMS's /admin/ page). We need to check if these files actually exist, which means they need to be loaded from the server directly, not from the cache - that's the purpose of this query, to prevent the cache handling them.

KyleAMathews commented on packages/gatsby/cache-dir/loader.js:
```diff
@@ -352,6 +319,12 @@ const queue = {
       if (!page) {
         console.log(`A page wasn't found for "${path}"`)
+
+        // Preload the custom 404 page when running gatsby-dev
```
*running gatsby develop

davidbailey00 replied:
Ah yep, I'll fix that

KyleAMathews commented on packages/gatsby/cache-dir/register-service-worker.js:
```diff
@@ -23,8 +23,14 @@ if (`serviceWorker` in navigator) {
-                apiRunner(`onServiceWorkerInstalled`, { serviceWorker: reg })
+                // Post to service worker that install is complete.
+                // Delay to allow time for the event listener to be added --
+                // otherwise fetch is called too soon and resources aren't cached.
```
oh interesting!

KyleAMathews commented on packages/gatsby/cache-dir/load-directly-or-404.js:
```diff
@@ -0,0 +1,39 @@
+  return new Promise(resolve => {
+    const url = new URL(window.location.origin + path)
+
+    // Check the page isn't already loaded directly.
```
How would we get here if the page had already loaded?

KyleAMathews commented:
Fantastic + through work! Really appreciate you diving into this. Been a sore spot for me that this was as broken as it was :-D

Thoughts on this btw? https://github.com/gatsbyjs/gatsby/pull/7355/commits/63ca8318eb114c521a776e70c6de3782f811e2a9

It seemed like a leftover so please PR it back if it is needed!

Merging and publishing!

KyleAMathews commented:
Get some well-deserved swag!

davidbailey00 commented:
@KyleAMathews Thanks for reviewing this!! I'll go and order myself some swag!

Re: 63ca831 I think that at some point while I was working on the fixes, without the check it was causing infinite reloads since it was constantly adding &no-cache=1 to the url, but it's safe to remove now given that the other problems which were causing that have been fixed - as you said, how could we get to those lines if the page had been loaded directly?

davidbailey00 commented:
@Dfmarulanda Please could you provide a repository which demonstrates the issue? There have been one or two reports of this but no reproduction repos as far as I know.

Also, it might just be that you need to clear your cache for the new service worker to install - you might still have the old one installed in your browser without the changes, even if you've updated the plugin on your website. Can you try this first and let me know if it works?

Dfmarulanda commented:
I already cleared the cache, updated, etc... The issue only happens when I'm an using Azure Web App. I invited you to my repo to check it out.

davidbailey00 commented:
@Dfmarulanda Thanks for the link, I'll check this out later and see if I can figure out what's wrong :)

davidbailey00 commented:
@Dfmarulanda The problem is that you're using an old version of gatsby-plugin-offline - shall I update it for you? Also, I see you've gitignored your lockfiles which is [not recommended](https://github.com/yarnpkg/yarn/issues/1583).

Dfmarulanda commented:
OK. Now is working, but i had to install de gatsby-plugin-offline. Without it, it's a never endless loop.

ssmith-wombatweb commented:
I'm having this same issue. When you go directly to my blog page it redirects to add a ?no-cache=1 and infinitely refreshes. It also happens if you refresh the page. Everything works fine if you go to it through the site links from another page. 

The blog page can be see at https://www.simeonsmith.me/coding-with-kids/2018/08/10/Crab-Trap-Alpha.

My repo is here: https://github.com/ssmith-wombatweb/simeonsmith.me

davidbailey00 commented:
@Dfmarulanda I just tried locally building your site with the latest version of gatsby-plugin-offline, and there are no problems. I then tried removing it, and there are still no problems - if there are still issues I think they must be caused by your Azure config, since it's working perfectly with local production builds. Perhaps it's because your `public` folder isn't gitignored and Azure isn't removing it properly pre-build? It's not the best idea to commit that folder because it's auto-generated and not useful to the build system which will just overwrite it.

davidbailey00 commented:
@ssmith-wombatweb So I think the lowercase rewrites are happening server-side, because it still gets rewritten when JavaScript is disabled (but of course it doesn't keep trying to reload without JS). Have you enabled Netlify's pretty URLs? Perhaps this is causing the issue.

davidbailey00 commented:
@ssmith-wombatweb ah, this must be a platform-specific bug since folders are case-insensitive on Windows, whereas on Unix environments it's working fine. I'll file a bug report later

ssmith-wombatweb commented:
I'm developing on a Mac and the reload is happening on both Windows and Mac. Does Netlify run on Windows?

davidbailey00 commented:
@ssmith-wombatweb Sorry, I mixed you up with Dfmarulanda - that seems strange. I'll do some testing later and see if I can figure out what's going wrong

davidbailey00 commented:
@ssmith-wombatweb Which branch are you working on at the moment, and deploying to Netlify? I've just tested with the master branch and it's working perfectly locally on Ubuntu and on Netlify (https://boring-hugle-14ed7e.netlify.com/coding-with-kids/2018/08/10/Crab-Trap-Alpha).

ssmith-wombatweb commented:
@davidbailey00 I am using the master branch. And when I follow that link I still get the reload issue. Did you clear the service worker cache and then visit that URL? Because I noticed if you visit other URLs at the site first and the go directly that link it doesn't have an issue. But if you go directly to that link without going to any other urls first it glitches.

To try and be as specific as possible it happens:
If you visit the blog article URL when it's the first time visiting the site and then anytime after that when you visit directly or refresh even if you visit other locations..
If you visit the blog article after clearing the service worker cache and also anytime after that when you visit directly or refresh even if you visit other locations.

It does not happen:
If you get there through the site, visiting other locations first and you haven't ever cleared the service worker cache.

This is an issue because:
If I want to share a link directly to my blog article it glitches for people who are going directly to the article and have never visited the site before.. 
It glitches for me because I occasionally need to clear the service worker cache to see the changes.

ssmith-wombatweb commented:
Another interesting note: 
If I have never cleared the service worker cache the first time I visit that URL directly it glitches. But if I visit another page and then go directly to that page it does not glitch.

davidbailey00 commented:
Ah sorry, I now see that the problem is still occurring. I'll have another look and see what I can find

jlengstorf commented:
@davidbailey00 would it make sense to make the `?no-cache` check case-insensitive? Seems like it could resolve this somewhat painlessly, right?

davidbailey00 commented:
I've figured this is definitely a problem with Netlify - it's rewriting uppercase URLs to lowercase ones, even with post processing disabled. Note that this is working locally (for me, at least) since the URL is kept uppercase, even though it's rewritten with a forward slash.

I recommend you [contact Netilfy's support](https://www.netlify.com/support/) and describe the problem, mentioning that Gatsby is case-sensitive and therefore the URL rewriting is causing problems - there should at least be an option to disable this in the build settings.

![image](https://user-images.githubusercontent.com/4248177/44808091-39b5fa80-abc3-11e8-9b75-20d1be1040fb.png)

ssmith-wombatweb commented:
Is it possible to have a Crab-Trap-Alpha and a crab-trap-alpha page? Not that I personally would do that but it seems like making that check case insensitive could cause issues there.

ssmith-wombatweb commented:
I was wondering that myself. Was gonna try and test it latter. Thank you so much for the help!

davidbailey00 commented:
@jlengstorf I could make it redirect to a lowercase version of the URL, but it would mean that all uppercase URLs are rewritten as lowercase whenever `?no-cache=1` is added, even when using a different host which doesn't have the same issue as Netlify.

Changing the behaviour in such a way would also cause edge cases if two pages exist with the same name except for capitalisation, so if I'm going to change this then I would also need to make Gatsby check at build time that there aren't any pages with the same name when lowercased.

davidbailey00 commented:
@ssmith-wombatweb I'd recommend that if you want to fix the issue quickly, you just rename the page so that it's all lowercase, while we figure out how to handle this best with Gatsby.

ssmith-wombatweb commented:
Thank you, that is my plan.

davidbailey00 commented:
@jlengstorf actually on second thoughts, I could modify `cache-dir/find-page.js` so that if a page isn't found at first, it checks again case-insensitively. That way, we wouldn't have to change the URL of the page, and it wouldn't affect hosts which serve URLs with the original capitalisation.

jlengstorf commented:
@davidbailey00 I'm _almost_ positive that URLs are case-insensitive. E.g. going to GOOGLE.COM and google.com lands us at the same place.

Am I missing some nuance here?

jlengstorf commented:
Quick non-Gatsby tests imply that this is true.

E.g. https://unsplash.com/ABOUT and https://unsplash.com/about hit the same page.

davidbailey00 commented:
@jlengstorf This is behaviour of the server - it's generally good for SEO to redirect different-case URLs to the same place, but they can be unique. Demo:

![image](https://user-images.githubusercontent.com/4248177/44812834-307f5a80-abd0-11e8-894d-74fa63775d1d.png)

davidbailey00 commented:
Also, with a different case:

![image](https://user-images.githubusercontent.com/4248177/44812919-6290bc80-abd0-11e8-9a62-2682bf2ebcd9.png)

jlengstorf commented:
I mean, I'd personally advocate for a "look for case collisions and error out with a 'this is terrible UX'" message over supporting an anti-pattern in Gatsby URLs. I can write up this doc and generate a shortlink for the error message if we decide this is the right path.

jlengstorf commented:
Yeah, I would agree that this is probably Gatsby core, not the offline plugin. @m-allanson @pieh @KyleAMathews can you weigh in here? This seems like something that would be good to sneak into v2, since it's technically a breaking change.

KyleAMathews commented:
URLs are case sensitive per the various internet RFCs for this https://stackoverflow.com/questions/7996919/should-url-be-case-sensitive/7996997

This doesn't seem like something we should fix as we could accidentally break sites where people care about case sensitivity. Dunno why Netlify ignores case but we can just tell people using Netlify to do everything lower case or whatever.

jlengstorf commented:
Those seem to imply that the server gets to make the call, so _users_ should expect case sensitivity.

To me, catching duplicate URLs with case differences is a UX improvement (e.g. `/About` and `/about` should be flagged because that's confusing and will most likely be accidental), and it wouldn't break any sites with case-sensitive URLs (e.g. both `/about` and `/About` still resolve to the about page). The only edge case would be developers that had deliberately put different pages under the same path with different cases, which to me sounds like an anti-pattern that we shouldn't support anyways (bad UX). (It also [reminds me of this comic](https://xkcd.com/1172/).)

The bugs we're seeing here seem far more likely to occur than someone deliberately obfuscating URLs, so I'd wager that the upside significantly outweighs the downside on this. Other hosting companies beyond Netlify coerce to lowercase, so this may be a recurring issue if we don't handle it inside Gatsby.

Alternatively, we could issue a warning for uppercase slugs: "Some hosting providers force URLs to lowercase. Using uppercase letters can break your site. You've been warned." That's lower effort, but also feels lower impact. 🤷‍♂️

KyleAMathews commented:
It seems weird to me to not support case sensitivity... perhaps we just add to our console output when we can't find a page that is sometimes due to case cohercion? And link to a docs page.

I agree there's probably not very many good use cases for case sensitivity but it also seems like the wrong call to try fix this edge case with a solution that potentially blocks off use cases we're not aware of as well as not supporting standard internet behavior.

This is a pretty unusual edge case and I think our best bet is to just do nothing and expect site builders to handle things.

davidbailey00 commented:
Perhaps this would be a good solution after all? It wouldn't break sites which have the edge case, but it would also fix problems with Netlfy / other hosts which rewrite URLs as lowercase.

![image](https://user-images.githubusercontent.com/4248177/44815180-b8686300-abd6-11e8-8285-7d15ae94f9b5.png)

jlengstorf commented:
Yeah, I'm not going to pick this hill to die on, so I think @davidbailey00's solution makes sense without making any policy decisions.

KyleAMathews commented:
Perhaps a focused warning but I'm wary of guessing the right thing for people here. It's easy enough for people to fix in their sites. If they're messing up casing when creating sites, they're probably messing up other things.

KyleAMathews commented:
Perhaps we'll make this change in the future but in general, it's unwise to make decisions with potentially far reaching effects based on one bug report. If it turns out this happens a lot we can reconsider.

ssmith-wombatweb commented:
Personally as someone who made this mistake and just had to learn about it I would have loved a warning. It's not something I've thought about before and being warned that there was a better way to do it is something I appreciate.

I personally like the option to go against convention but do appreciate being told about it.

Dfmarulanda commented:
It wasn't a problem of azure, because it is a server as a static website. The problem only happened when in the production build served by Azure. It never happened on the local build. @davidbailey00 I'm uploading the public folder because I use it to make a continuous deployment using my GitHub repo. It's fixed, but it was a weird bug generated in the last update.

davidbailey00 commented:
@KyleAMathews @ssmith-wombatweb @Dfmarulanda @jlengstorf  perhaps we could have a runtime console warning after `A page wasn't found for ...` if it detects a page with the same name but different casing? We won't fix the Netlify+uppercase edge case but then it should be clear why it's not working.

On the note of this, I just noticed that URL search queries aren't always kept when redirecting to `?no-cache=1` - given the Netlify problem, https://www.simeonsmith.me/coding-with-kids/2018/08/10/crab-trap-alpha?no-cache=1 should keep adding `&no-cache=1` when reloading. A better demonstration of the problem:

1. Go to https://wizardly-babbage-0faa32.netlify.com/ and wait for the SW to install

2. Then go to https://wizardly-babbage-0faa32.netlify.com/admin/?somequery=1 (should be handled by the service worker)

3. Expected behaviour: redirects to https://wizardly-babbage-0faa32.netlify.com/admin/?somequery=1&no-cache=1

    Actual behaviour: redirects to https://wizardly-babbage-0faa32.netlify.com/admin/?no-cache=1

davidbailey00 commented:
@KyleAMathews Alright, I'm still doing work experience until the end of this week but I can try fixing that tomorrow evening, along with the other bug.

A warning should be easy to implement, although the bug I just discovered might take me a bit longer - it only happens from external links or direct URL entry, not from `<Link>`s; I think I vaguely understand why but I'll need to play around a bit. I can submit these as separate PRs since they're not directly related, unlike the issues from #7355.

(Also, unrelated to these issues but the company I'm doing work exp at seems impressed by Gatsby and is considering rewriting their current sites on legacy stacks using it :smile:)

ssmith-wombatweb commented:
Just coming back to confirm that changing that to lowercase solves my issues on Netlify.
https://www.simeonsmith.me/coding-with-kids/2018/08/10/crab-trap-alpha

davidbailey00 commented:
I've thought of a better solution than showing a warning - we can use Gatsby's global `window.page.path` to get the canonical path rather than using the page URL, which means that even if the page is loaded from a different location (potentially completely different), it can still find the resources. I'lll work on this tonight

KyleAMathews commented:
>(Also, unrelated to these issues but the company I'm doing work exp at seems impressed by Gatsby and is considering rewriting their current sites on legacy stacks using it 😄)

Nice! :-D

davidbailey00 commented:
> I've thought of a better solution than showing a warning - we can use Gatsby's global window.page.path to get the canonical path rather than using the page URL, which means that even if the page is loaded from a different location (potentially completely different), it can still find the resources. I'lll work on this tonight.

Just to update everyone, I've made some progress on this by replacing `window.location.pathname` with `window.page.path` in `production-app.js` but currently trying to understand how the `<PageRenderer>`'s properties are set, since modifying just `production-app.js` doesn't fix the issue.

davidbailey00 commented:
cc @KyleAMathews @jlengstorf @m-allanson I'm really struggling to understand where `RouteHandler` gets its props from, can someone help me out? I'm testing by making a copy of the `about` folder to `about-test`, to see if I can get it to load correctly using the canonical path (`window.page.path`) rather than the URL's actual path.

If I log the props...

![image](https://user-images.githubusercontent.com/4248177/44881952-4a40a080-aca9-11e8-8f58-35c492120e4c.png)

...it's clear that the `location` prop is set from `window.location`:

![image](https://user-images.githubusercontent.com/4248177/44881982-63495180-aca9-11e8-8a1e-971e50b3bd8a.png)

But after running a trace I still have no idea how it gets these props. The element itself is created at line 126 (in my local code with added `console.log`s) when the `renderer` function is called:

![image](https://user-images.githubusercontent.com/4248177/44882075-bf13da80-aca9-11e8-8f15-99b32d6dc1d8.png)

...since it renders `NewRoute`, which wraps the `RouteHandler` element. It's clear that the only prop passed is the `path` prop:

![image](https://user-images.githubusercontent.com/4248177/44882113-e074c680-aca9-11e8-9e60-1d16657bd380.png)

All function calls between line 126 and the `RouteHandler` constructor are React calls, which can be seen via the console trace:

![image](https://user-images.githubusercontent.com/4248177/44882211-46f9e480-acaa-11e8-9f37-24f524427682.png)

All the lower calls are higher-level and have no mention of `location` or anything like that. My understanding is that Reach Router must be adding the props, but I don't understand it well enough to know how this works... Could someone help me out please?

KyleAMathews commented:
@davidbailey00 could you copy your previous comments over to a new issue? This one is getting rather lengthy :-)

Dfmarulanda commented:
I updated gatsby today and now im getting 

```terminal
These dependencies were not found:

* @babel/runtime/core-js/get-iterator in ./node_modules/gatsby-plugin-offline/gatsby-browser.js
* @babel/runtime/core-js/promise in ./node_modules/gatsby-plugin-offline/gatsby-browser.js
```

davidbailey00 commented:
@Dfmarulanda Thanks for the heads up, I'll check this out.

davidbailey00 commented:
@Dfmarulanda hm, I'm unable to see what the issue is here since everything works locally for me... Can you try upgrading all packages again, and running `yarn` (or `npm install`) just to make sure any new dependencies are installed? If that still fails, can you output your `package.json`?

Dfmarulanda commented:
I already deleted node_modules and then yarn install. Also tried to add @babel/runtime by itself.

```json 
{
  "private": true,
  "scripts": {
    "build": "gatsby build",
    "develop": "gatsby develop",
    "lint:es": "eslint --ignore-path .gitignore \"**/*.{js,jsx}\"",
    "lint:style": "stylelint \"src/**/*.scss\"",
    "lint": "yarn lint:es && yarn lint:style",
    "format": "prettier-eslint --write \"**/*.{js,jsx,json,css,scss,md}\""
  },
  "dependencies": {
    "@babel/runtime": "^7.0.0",
    "@reach/router": "^1.1.1",
    "babel-eslint": "^8.2.6",
    "babel-plugin-styled-components": "^1.5.1",
    "fbjs": "^0.8.17",
    "gatsby": "^2.0.0-rc.7",
    "gatsby-plugin-canonical-urls": "^2.0.0-rc.1",
    "gatsby-plugin-drift": "^1.0.0",
    "gatsby-plugin-favicon": "^3.1.2",
    "gatsby-plugin-offline": "2.0.0-rc.0",
    "gatsby-plugin-react-helmet": "^3.0.0-rc.1",
    "gatsby-plugin-styled-components": "^3.0.0-rc.1",
    "gatsby-plugin-typography": "^2.2.0-rc.2",
    "grid-styled": "^5.0.2",
    "object-assign": "^4.1.1",
    "prop-types": "^15.6.2",
    "react": "^16.4.2",
    "react-dom": "^16.4.2",
    "react-emotion": "^9.2.4",
    "react-helmet": "^5.2.0",
    "react-router-dom": "^4.3.1",
    "react-typography": "^0.16.13",
    "styled-components": "^3.4.5",
    "styled-media-query": "^2.0.2",
    "typography": "^0.16.17"
  },
  "devDependencies": {
    "emotion": "^9.2.6",
    "eslint": "^4.19.1",
    "eslint-config-airbnb": "^17.0.0",
    "eslint-plugin-import": "^2.13.0",
    "eslint-plugin-jsx-a11y": "^6.1.1",
    "eslint-plugin-react": "^7.10.0",
    "prettier-eslint-cli": "^4.7.0",
    "stylelint": "^9.4.0",
    "stylelint-config-idiomatic-order": "^5.0.0",
    "stylelint-config-recommended-scss": "^3.0.0",
    "stylelint-scss": "^3.2.0",
    "webpack": "^4.16.4",
    "webpack-cli": "^3.1.0"
  }
}
```

davidbailey00 commented:
You've got an old version of `gatsby-plugin-offline` which isn't updating because for some reason it doesn't have the caret (I can see that all other packages do). rc2 is the latest version; can you add the caret and then upgrade again?

![image](https://user-images.githubusercontent.com/4248177/44935046-1e83f000-ad67-11e8-8543-89fc448cd804.png)

Dfmarulanda commented:
Yeah. It seems it was the problem. Thanks!

Dfmarulanda commented:
I made an update yesterday and it seems it's broken again. I cannot figure out what's happening with it. You can check out my website at https://crowdswap.com

pieh commented:
@Dfmarulanda can you create new issue with more details of your setup - including output from `gatsby info`?

davidbailey00 commented:
@pieh actually I'm just checking this out and the problem seems to be because the `public` folder isn't gitignored, so might not need a new issue

davidbailey00 commented:
@Dfmarulanda I've just checked and confirmed my above comment - you're not ignoring the `public` folder so the wrong files are being generated, causing additional errors. Please add `public/` to your gitignore and then delete the folder, and commit all changes. Working demo: https://clever-villani-54e8ff.netlify.com/

Dfmarulanda commented:
But i don't understand why i have to gitignore the public folder. I use it was continous deployment to my website using azure web apps. I always delete the public folder before publishing. Why it would be the problem?

davidbailey00 commented:
@Dfmarulanda Your HTML file is referencing JSON files which don't exist on your server:

Dfmarulanda commented:
They exist, I checked those files multiple times.

Dfmarulanda commented:
OK. I'll try to change my deploy process. It's super easy to do it using netlify, aws, heroku, but there is no an azure plugin. There is a docker plugin working in the latest version of Gatsby?
